base/applications/mspaint/mouse.cpp

```diff
@@ -277,293 +508,374 @@ endPaintingL(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
+    {
     }
 
-    last.x = x;
-    last.y = y;
-}
+    void draw(BOOL bLeftButton, LONG x, LONG y)
+    {
+        if (bLeftButton)
+            Rect(m_hdc, start.x, start.y, x, y, m_fg, m_bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+        else
+            Rect(m_hdc, start.x, start.y, x, y, m_bg, m_fg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle());
+    }
 
-void
-endPaintingR(HDC hdc, LONG x, LONG y, COLORREF fg, COLORREF bg)
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        imageModel.CopyPrevious();
+    }
+
+    void OnMouseMove(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnMouseMove(bLeftButton, x, y);
+    }
+
+    void OnButtonUp(BOOL bLeftButton, LONG x, LONG y)
+    {
+        imageModel.ResetToPrevious();
+        if (GetAsyncKeyState(VK_SHIFT) < 0)
+            regularize(start.x, start.y, x, y);
+        draw(bLeftButton, x, y);
+        ToolBase::OnButtonUp(bLeftButton, x, y);
+    }
+
+    void OnCancelDraw()
+    {
+        OnButtonUp(FALSE, 0, 0);
+        imageModel.Undo();
+        selectionModel.ResetPtStack();
+        ToolBase::OnCancelDraw();
+    }
+};
+
+// TOOL_SHAPE
+struct ShapeTool : ToolBase
 {
-    switch (toolsModel.GetActiveTool())
-    {
-        case TOOL_RUBBER:
-            Replace(hdc, last.x, last.y, x, y, fg, bg, toolsModel.GetRubberRadius());
-            break;
-        case TOOL_PEN:
-            Line(hdc, last.x, last.y, x, y, bg, 1);
-            SetPixel(hdc, x, y, bg);
-            break;
-        case TOOL_LINE:
-            imageModel.ResetToPrevious();
-            if (GetAsyncKeyState(VK_SHIFT) < 0)
-                roundTo8Directions(start.x, start.y, x, y);
-            Line(hdc, start.x, start.y, x, y, bg, toolsModel.GetLineWidth());
-            break;
-        case TOOL_BEZIER:
+    ShapeTool() : ToolBase(TOOL_SHAPE)
+    {
+    }
+
+    void draw(BOOL bLeftButton, LONG x, LONG y, BOOL bClosed = FALSE)
+    {
+        if (bLeftButton)
+        {
+            if (pointSP + 1 >= 2)
+                Poly(m_hdc, pointStack, pointSP + 1, m_fg, m_bg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle(), bClosed, FALSE);
+        }
+        else
+        {
+            if (pointSP + 1 >= 2)
+                Poly(m_hdc, pointStack, pointSP + 1, m_bg, m_fg, toolsModel.GetLineWidth(), toolsModel.GetShapeStyle(), bClosed, FALSE);
+        }
+    }
+
+    void OnButtonDown(BOOL bLeftButton, LONG x, LONG y, BOOL bDoubleClick)
+    {
+        ToolBase::OnButtonDown(bLeftButton, x, y, bDoubleClick);
+        pointStack[pointSP].x = x;
+        pointStack[pointSP].y = y;
+        if (pointSP == 0)
+        {
+            imageModel.CopyPrevious();
+            draw(bLeftButton, x, y);
```

Comment:
Question: Is this a fix, compared to the old code?
Old code always did the `Poly` call (what your `draw()` function now does), and then, **only if `pointSP == 0`**, did it `CopyPrevious` . While I agree it's better to do that latter call first, shouldn't the `draw()` call be always done?

Because, note also that since now you call `draw` only when `pointSP == 0`, the conditions `if (pointSP + 1 >= 2)` in `draw` would never be TRUE.

Reply:
Fixed.